src/persist/src/file.rs

```diff
@@ -0,0 +1,207 @@
+            let ts = u64::from_le_bytes(buf);
+
+            // Decode diff
+            bytes.read_exact(&mut buf)?;
+            let diff = i64::from_le_bytes(buf);
+
+            dataz.push(((key, val), ts, diff as isize));
+        }
+        Ok(MemSnapshot { dataz })
+    }
+
+    fn allow_compaction(&mut self, _ts: u64) -> Result<(), Error> {
+        // No-op for now.
+        Ok(())
+    }
+}
+
+/// Implementation of [Persister] backed by files.
+pub struct FilePersister {
+    base_path: PathBuf,
+    registered: HashSet<Id>,
+    dataz: HashMap<Id, FileStream>,
+}
+
+impl FilePersister {
+    /// Constructs a new, empty FilePersister with the specified base path.
+    pub fn new<P: AsRef<Path>>(base_path: P) -> Result<Self, Error> {
+        fs::create_dir_all(&base_path)?;
+        Ok(FilePersister {
+            base_path: base_path.as_ref().to_path_buf(),
+            registered: HashSet::new(),
+            dataz: HashMap::new(),
+        })
+    }
+}
+
+impl Persister for FilePersister {
+    type Write = FileStream;
+    type Meta = FileStream;
+
+    fn create_or_load(&mut self, id: Id) -> Result<Token<Self::Write, Self::Meta>, Error> {
+        if self.registered.contains(&id) {
+            return Err(format!("internal error: {:?} already registered", id).into());
+        }
+
+        self.registered.insert(id);
+        let path = self.base_path.join(format!("file-persister-{}", id.0));
+        let file_stream = FileStream::new(path)?;
+        let p = self.dataz.entry(id).or_insert(file_stream);
+        let t = Token {
+            write: p.clone(),
+            meta: p.clone(),
+        };
+        Ok(t)
+    }
+
+    /// Removes the persisted stream for `id`.
+    ///
+    /// Note that this function does not unregister `id` to ensure that each `id`
+    /// is only ever registered exactly once.
+    fn destroy(&mut self, id: Id) -> Result<(), Error> {
+        if self.dataz.remove(&id).is_none() {
+            return Err(format!("internal error: {:?} not registered", id).into());
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::{SystemTime, UNIX_EPOCH};
+
+    use super::*;
+
+    #[test]
+    fn test_file_stream() {
+        let data = vec![
+            (("key1".to_string(), "val1".to_string()), 1, 1),
+            (("key2".to_string(), "val2".to_string()), 1, 1),
+        ];
+
+        // Create a directory that will automatically be dropped after the test finishes.
+        let temp_dir = tempfile::tempdir().expect("failed to create temp directory");
+        let file_path = temp_dir.path().join(format!(
+            "test-{}",
+            SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("clock went backwards")
+                .as_nanos()
+        ));
+        let mut file_stream = FileStream::new(file_path).expect("failed to create file stream");
+        file_stream.write_sync(&data).expect("fixme");
```

Comment:
tests can return result, so lets do that instead of this expect :)

(i think the benchmark can't return result tho, so those ones will have to stay)